filecoin-proofs/src/api.rs

```diff
@@ -67,38 +67,230 @@ pub struct SealOutput {
+    }
+}
+
+impl std::cmp::PartialOrd for ReplicaInfo {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl ReplicaInfo {
+    pub fn new(access: String, commitment: Commitment) -> Self {
+        ReplicaInfo { access, commitment }
+    }
+
+    pub fn safe_commitment(&self) -> Result<PedersenDomain, failure::Error> {
+        as_safe_commitment(&self.commitment, "comm_r")
+    }
+
+    /// Generate the merkle tree of this particular replica.
+    pub fn merkle_tree(&self, sector_size: u64) -> Result<Tree, Error> {
+        let mut f_in = File::open(&self.access)?;
+        let mut data = Vec::new();
+        f_in.read_to_end(&mut data)?;
+
+        let bytes = PaddedBytesAmount(sector_size as u64);
+        public_params(bytes, 1).graph.merkle_tree(&data)
+    }
+}
+
+fn as_safe_commitment(
+    comm: &Commitment,
+    commitment_name: impl AsRef<str>,
+) -> Result<PedersenDomain, failure::Error> {
+    bytes_into_fr::<Bls12>(comm).map(Into::into).map_err(|err| {
+        format_err!(
+            "Invalid commitment ({}): {:?}",
+            commitment_name.as_ref(),
+            err
+        )
+    })
+}
+
+/// Generates a proof-of-spacetime.
 /// Accepts as input a challenge seed, configuration struct, and a vector of
-/// sealed sector file-path plus CommR tuples.
-///
+/// sealed sector file-path plus CommR tuples, as well as a list of faults.
 pub fn generate_post(
     post_config: PoStConfig,
-    challenge_seed: ChallengeSeed,
-    input_parts: Vec<(Option<String>, Commitment)>,
-) -> error::Result<GeneratePoStDynamicSectorsCountOutput> {
-    generate_post_dynamic(GeneratePoStDynamicSectorsCountInput {
-        post_config,
+    challenge_seed: &ChallengeSeed,
+    replicas: Vec<ReplicaInfo>,
+    faults: &[u64],
+) -> error::Result<Vec<u8>> {
+    let sector_count = replicas.len() as u64;
+    let sector_size = u64::from(PaddedBytesAmount::from(post_config));
+
+    // Only one fault per sector is possible.
+    ensure!(
+        faults.len() as u64 <= sector_count,
+        "Too many faults submitted"
+    );
+
+    let vanilla_params = post_setup_params(post_config);
+    let setup_params = compound_proof::SetupParams {
+        vanilla_params: &vanilla_params,
+        engine_params: &(*ENGINE_PARAMS),
+        partitions: None,
+    };
+
+    let pub_params: compound_proof::PublicParams<_, rational_post::RationalPoSt<PedersenHasher>> =
+        RationalPoStCompound::setup(&setup_params)?;
+
+    let challenges = rational_post::derive_challenges(
+        vanilla_params.challenges_count,
+        sector_size,
+        sector_count,
         challenge_seed,
-        input_parts,
-    })
+        &faults,
+    );
+
+    // Match the replicas to the challenges, as these are the only ones required.
+    let challenged_replicas: Vec<_> = challenges
+        .iter()
+        .map(|c| {
+            if let Some(replica) = replicas.get(c.sector as usize) {
```

Comment:
BLOCKING: `replicas.get(c.sector as usize)` is going to index into `replicas` by sector id. `replicas` is a `Vec` (i.e. not sector id-indexed data structure).

Reply:
How do I get the sector id from the the replica then?

Reply:
You'll need to extend `ReplicaInfo` to include this information, e.g.

```
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct ReplicaInfo {
    /// Path to the replica.
    access: String,
    /// The replica commitment.
    commitment: Commitment,

    sector_id: SectorId
}
```

Reply:
We should pause this thread until we're all on the same page regarding whether or not `GeneratePoSt` should accept a `SectorSet` (a sector id-indexed map) or an ordered list of sector ids.